### DISCUSSION TITLE
Add ability to select sketch regions for extrude, revolve, etc

### DISCUSSION THREAD

**Discussion by franknoirot** [KittyCAD/modeling-app#7494] [Milestones]
> [!IMPORTANT]
> This is currently in planning, see discussion below.
> This is a sister milestone to #7435, see discussions there as well.

Sketch regions are a commonly-used concept in CAD modeling. A sketch may contain many closed or open profiles, and a sketch region could be defined as the intersection of two or more closed profiles. Supporting this concept allows users to, for example, sketch a plate and the holes they want in its corners in the same sketch, and then when extruding, exclude the holes in the extruded sketch region.

## TradCAD demos

Note in these demos how it is not the entirety of the outer profile that is selected for extrusion, but rather the _region_ of the outer profile _excluding_ any inner profiles. The user can then opt into including any of the inner profiles by clicking inside them afterward to add them to the selection.

https://github.com/user-attachments/assets/4a17144a-8b73-42fa-aaa7-351d671195a3

## Engine notes
1. None, regions are not a concept in-engine. At some point it may be good to have the engine support that concept directly, but initial implementation may not need it.
2. Previous discussion about the "hole" approach discussed below resulted in an endpoint to detect the profiles that lie within an outer one, which can be found [here](https://github.com/KittyCAD/engine/pull/2351)
## API/KCL notes
A region is somewhat representable using a pipe expression from the outermost profile to any inner profiles to be excluded, wrapped with our `hole` function. One limitation of this is our [existing behavior when holed inner profiles intersect the outer profile](https://github.com/KittyCAD/engine/issues/1385), since it naively inverts the extrusion and does not exclude the portion lying outside the outer profile.

Here is a video explanation of this hole approach being used with hand-written KCL. A discussion of how this might be automatically done for the user in point-and-click is below.

https://github.com/user-attachments/assets/224de244-a39c-44e4-aa89-c293aa67eaab

## Point-and-click notes
1. Sketch regions are not created unless needed by a 3D operation
2. They are marked as needing to be created based on the selection argument step in the command bar
3. We hit the engine endpoint to get info of the selection location about inner profiles
4. If any exist, we mark the code mod as needing to generate region, and the outer and inner profiles
    1. I don't believe we have a way to show the user a corresponding visual of these regions until we have an engine-side representation of regions
5. On operation submission, the code mod will produce a region definition that is a pipe expression of the outer profile and inner profiles holed out, and use that region for the operation instead of the whole sketch

**Comment by max-mrgrsk** [comment]
first idea:
 - on the point and click side we can make it identical to others (you are free to select any region(s)).
 - on the kcl side code mod automatically construct the intersect+subtract+union chains for each region in the selection. 

simple case is almost there, the missing things are intersect and union. And point and click support ofc
<img width="1468" alt="Bildschirmfoto 2025-06-16 um 16 50 26" src="https://github.com/user-attachments/assets/766c9ae1-99db-417f-8a21-57e008ba6090" />

**Reply by franknoirot** [reply]
I like this! `subtract2d` needs to be made more robust for this; the legacy behavior from when it was called `hole` is insufficient to support this use case, as [partial overlaps are not handled correctly by subract2d at the moment](https://github.com/kittycad/engine/issues/1385).

**Reply by max-mrgrsk** [reply]
cool, totally agree!
here is more complex part of selecting several regions within one selection (for further extrude, revolve etc.). issues:
- array of regions
- chaining region operations 
- cases when you have several regions out of one kcl. same problem as with csg we have discussed half a year ago. 


<img width="1553" alt="Bildschirmfoto 2025-06-16 um 17 15 22" src="https://github.com/user-attachments/assets/99ee3394-0409-4cef-86f6-0b14fa3e7efb" />

**Reply by max-mrgrsk** [reply]
@franknoirot link does not work for some reason

**Reply by pierremtb** [reply]
@max-mrgrsk I believe he meant https://github.com/kittycad/engine/issues/1385

**Reply by franknoirot** [reply]
Thanks @pierremtb yup yup, fixed my URL

**Comment by Irev-Dev** [comment]
Notes getting scattered a little, I did have a proposal that avoids "2d CSG" here
https://github.com/KittyCAD/modeling-app/discussions/7435#discussioncomment-13494803

Btw I think the problem  with 2D csg is it requires all of the profiles to be closed for it to work
i.e. this
![image](https://github.com/user-attachments/assets/b51daa91-adf0-4db8-aaa1-53f3f52b30fe)

**Comment by pierremtb** [comment]
I believe we should close this discussion, and instead have it as a requirement of #7435

**Reply by pierremtb** [reply]
@franknoirot @Irev-Dev lmk if you disagree
